src/module_wrap.cc

```diff
@@ -564,22 +564,20 @@ static MaybeLocal<Promise> ImportModuleDynamically(
 
   Local<Function> import_callback =
     env->host_import_module_dynamically_callback();
+  Local<Value> id;
 
   Local<FixedArray> options = host_defined_options.As<FixedArray>();
-  if (options->Length() != HostDefinedOptions::kLength) {
-    Local<Promise::Resolver> resolver;
-    if (!Promise::Resolver::New(context).ToLocal(&resolver)) return {};
-    resolver
-        ->Reject(context,
-                 v8::Exception::TypeError(FIXED_ONE_BYTE_STRING(
-                     context->GetIsolate(), "Invalid host defined options")))
-        .ToChecked();
-    return handle_scope.Escape(resolver->GetPromise());
+  // Get referrer id symbol from the host-defined options.
+  // If the host-defined options are empty, get the referrer id symbol
+  // from the realm global object.
+  if (options->Length() == HostDefinedOptions::kLength) {
+    id = options->Get(context, HostDefinedOptions::kID).As<Symbol>();
+  } else {
+    id = context->Global()
```

Comment:
Doesn't this abort now in the case of https://github.com/nodejs/node/issues/49726 ? I think we should either throw as before, or do a fallback for the main context via the symbol (but what the referrer should be in the fallback is still pending resolution on the spec side, so I think throwing as before is a better intermediate situation).

Reply:
As mentioned in the PR description, this PR opens up the possibility to define a custom handler for the case when the referrer is a realm. Libraries can determine their own resolution in this valid case.

However, the resolution with a realm referrer in the main context is not settled yet, and the spec might change to remove the propagation requirement of the active script or module to reflect the web reality. So this PR doesn't change the behavior described in #49726 (well, the thrown error is changed: it is not "Invalid host defined options" but "ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING" now.)

I can add a test to ensure that this behavior is not easily changed.